{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SpatialDE\n",
    "\n",
    "SpatialDE is a method to identify genes which signfificantly depend on spatial coordinates in non-linear and non-parametric ways. The intended applications are spatially resolved RNA-sequencing from e.g. Spatial Transcriptomics, or _in situ_ gene expression measurements from e.g. SeqFISH or MERFISH.\n",
    "\n",
    "This repository contains both the implementations of our method, as well as case studies in applying it.\n",
    "\n",
    "The key features of our method are\n",
    "\n",
    "  - Unsupervised - No need to define spatial regions\n",
    "  - Non-parametric and non-linear expression patterns\n",
    "  - Extremely fast - Transcriptome wide tests takes only a few minutes on normal computers\n",
    "  \n",
    "The primary implementation is as a Python3 package, and can be installed from the command line by \n",
    "\n",
    "```\n",
    "$ pip install spatialde\n",
    "```\n",
    "\n",
    "Below follows a typical usage example in interactive form."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    }
   ],
   "source": [
    "%pylab inline\n",
    "import pandas as pd\n",
    "\n",
    "rcParams['axes.spines.right'] = False\n",
    "rcParams['axes.spines.top'] = False\n",
    "\n",
    "import NaiveDE\n",
    "import SpatialDE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Nrf1</th>\n",
       "      <th>Zbtb5</th>\n",
       "      <th>Ccnl1</th>\n",
       "      <th>Lrrfip1</th>\n",
       "      <th>Bbs1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16.92x9.015</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16.945x11.075</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16.97x10.118</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16.939x12.132</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16.949x13.055</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Nrf1  Zbtb5  Ccnl1  Lrrfip1  Bbs1\n",
       "16.92x9.015       1      1      1        2     1\n",
       "16.945x11.075     0      0      3        2     2\n",
       "16.97x10.118      0      1      1        0     0\n",
       "16.939x12.132     1      0      1        0     4\n",
       "16.949x13.055     0      0      0        3     0"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts = pd.read_csv('Analysis/MouseOB/data/Rep11_MOB_0.csv', index_col=0)\n",
    "counts = counts.T[counts.sum(0) >= 3].T  # Filter practically unobserved genes\n",
    "\n",
    "counts.iloc[:5, :5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>total_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16.92x9.015</th>\n",
       "      <td>16.920</td>\n",
       "      <td>9.015</td>\n",
       "      <td>18790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16.945x11.075</th>\n",
       "      <td>16.945</td>\n",
       "      <td>11.075</td>\n",
       "      <td>36990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16.97x10.118</th>\n",
       "      <td>16.970</td>\n",
       "      <td>10.118</td>\n",
       "      <td>12471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16.939x12.132</th>\n",
       "      <td>16.939</td>\n",
       "      <td>12.132</td>\n",
       "      <td>22703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16.949x13.055</th>\n",
       "      <td>16.949</td>\n",
       "      <td>13.055</td>\n",
       "      <td>18641</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    x       y  total_counts\n",
       "16.92x9.015    16.920   9.015         18790\n",
       "16.945x11.075  16.945  11.075         36990\n",
       "16.97x10.118   16.970  10.118         12471\n",
       "16.939x12.132  16.939  12.132         22703\n",
       "16.949x13.055  16.949  13.055         18641"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_info = pd.read_csv('Analysis/MouseOB/MOB_sample_info.csv', index_col=0)\n",
    "counts = counts.loc[sample_info.index]  # Align count matrix with metadata table\n",
    "\n",
    "sample_info.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "norm_expr = NaiveDE.stabilize(counts.T).T\n",
    "resid_expr = NaiveDE.regress_out(sample_info, norm_expr.T, 'np.log(total_counts)').T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the sake of this example, let's just run the test on 1000 random genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Performing DE test\n",
      "INFO:root:Pre-calculating USU^T = K's ...\n",
      "INFO:root:Done: 0.076s\n",
      "INFO:root:Fitting gene models\n",
      "INFO:root:Model 1 of 10\n",
      "INFO:root:Model 2 of 10                             \n",
      "INFO:root:Model 3 of 10                            \n",
      "INFO:root:Model 4 of 10                             \n",
      "INFO:root:Model 5 of 10                            \n",
      "INFO:root:Model 6 of 10                             \n",
      "INFO:root:Model 7 of 10                             \n",
      "INFO:root:Model 8 of 10                            \n",
      "INFO:root:Model 9 of 10                             \n",
      "INFO:root:Model 10 of 10                            \n",
      "                                                    \r"
     ]
    }
   ],
   "source": [
    "sample_resid_expr = resid_expr.sample(n=1000, axis=1, random_state=24)\n",
    "\n",
    "X = sample_info[['x', 'y']]\n",
    "results = SpatialDE.run(X, sample_resid_expr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The result will be a DataFrame with P-values and other relevant values for each gene.\n",
    "\n",
    "The most important columns are\n",
    "\n",
    "  - `g` - The name of the gene\n",
    "  - `pval` - The P-value for spatial differential expression\n",
    "  - `qval` - Signifance after correcting for multiple testing\n",
    "  - `l` - A parameter indicating the distance scale a gene changes expression over"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gower</th>\n",
       "      <td>0.999295</td>\n",
       "      <td>0.999295</td>\n",
       "      <td>0.999295</td>\n",
       "      <td>0.999295</td>\n",
       "      <td>0.999295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g</th>\n",
       "      <td>Tinagl1</td>\n",
       "      <td>Vstm2l</td>\n",
       "      <td>6330415B21Rik</td>\n",
       "      <td>Galnt4</td>\n",
       "      <td>Leng8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>l</th>\n",
       "      <td>0.402001</td>\n",
       "      <td>0.402001</td>\n",
       "      <td>0.402001</td>\n",
       "      <td>0.402001</td>\n",
       "      <td>0.402001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max_delta</th>\n",
       "      <td>0.00628877</td>\n",
       "      <td>0.0484324</td>\n",
       "      <td>0.837928</td>\n",
       "      <td>0.00806104</td>\n",
       "      <td>0.975425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max_ll</th>\n",
       "      <td>11.5958</td>\n",
       "      <td>-125.505</td>\n",
       "      <td>232.757</td>\n",
       "      <td>91.4048</td>\n",
       "      <td>-87.1177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max_mu_hat</th>\n",
       "      <td>0.025265</td>\n",
       "      <td>-4.84373</td>\n",
       "      <td>0.877441</td>\n",
       "      <td>0.611605</td>\n",
       "      <td>-1.56887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max_s2_t_hat</th>\n",
       "      <td>0.0540122</td>\n",
       "      <td>19.2345</td>\n",
       "      <td>0.386566</td>\n",
       "      <td>0.34356</td>\n",
       "      <td>1.19909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <td>SE</td>\n",
       "      <td>SE</td>\n",
       "      <td>SE</td>\n",
       "      <td>SE</td>\n",
       "      <td>SE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n</th>\n",
       "      <td>260</td>\n",
       "      <td>260</td>\n",
       "      <td>260</td>\n",
       "      <td>260</td>\n",
       "      <td>260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>time</th>\n",
       "      <td>0.000803947</td>\n",
       "      <td>0.000534058</td>\n",
       "      <td>0.000344038</td>\n",
       "      <td>0.000543118</td>\n",
       "      <td>0.000329018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BIC</th>\n",
       "      <td>-0.948912</td>\n",
       "      <td>273.252</td>\n",
       "      <td>-443.272</td>\n",
       "      <td>-160.567</td>\n",
       "      <td>196.478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max_ll_null</th>\n",
       "      <td>10.697</td>\n",
       "      <td>-126.502</td>\n",
       "      <td>232.255</td>\n",
       "      <td>90.0269</td>\n",
       "      <td>-87.3105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LLR</th>\n",
       "      <td>0.898786</td>\n",
       "      <td>0.997314</td>\n",
       "      <td>0.502405</td>\n",
       "      <td>1.37795</td>\n",
       "      <td>0.192801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fraction_spatial_variance</th>\n",
       "      <td>0.993746</td>\n",
       "      <td>0.953774</td>\n",
       "      <td>0.543916</td>\n",
       "      <td>0.991998</td>\n",
       "      <td>0.506044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pval</th>\n",
       "      <td>0.343107</td>\n",
       "      <td>0.317961</td>\n",
       "      <td>0.478445</td>\n",
       "      <td>0.240451</td>\n",
       "      <td>0.660595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>qval</th>\n",
       "      <td>0.977568</td>\n",
       "      <td>0.977568</td>\n",
       "      <td>0.977568</td>\n",
       "      <td>0.977568</td>\n",
       "      <td>0.977568</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     0            1              2  \\\n",
       "Gower                         0.999295     0.999295       0.999295   \n",
       "M                                    4            4              4   \n",
       "g                              Tinagl1       Vstm2l  6330415B21Rik   \n",
       "l                             0.402001     0.402001       0.402001   \n",
       "max_delta                   0.00628877    0.0484324       0.837928   \n",
       "max_ll                         11.5958     -125.505        232.757   \n",
       "max_mu_hat                    0.025265     -4.84373       0.877441   \n",
       "max_s2_t_hat                 0.0540122      19.2345       0.386566   \n",
       "model                               SE           SE             SE   \n",
       "n                                  260          260            260   \n",
       "time                       0.000803947  0.000534058    0.000344038   \n",
       "BIC                          -0.948912      273.252       -443.272   \n",
       "max_ll_null                     10.697     -126.502        232.255   \n",
       "LLR                           0.898786     0.997314       0.502405   \n",
       "fraction_spatial_variance     0.993746     0.953774       0.543916   \n",
       "pval                          0.343107     0.317961       0.478445   \n",
       "qval                          0.977568     0.977568       0.977568   \n",
       "\n",
       "                                     3            4  \n",
       "Gower                         0.999295     0.999295  \n",
       "M                                    4            4  \n",
       "g                               Galnt4        Leng8  \n",
       "l                             0.402001     0.402001  \n",
       "max_delta                   0.00806104     0.975425  \n",
       "max_ll                         91.4048     -87.1177  \n",
       "max_mu_hat                    0.611605     -1.56887  \n",
       "max_s2_t_hat                   0.34356      1.19909  \n",
       "model                               SE           SE  \n",
       "n                                  260          260  \n",
       "time                       0.000543118  0.000329018  \n",
       "BIC                           -160.567      196.478  \n",
       "max_ll_null                    90.0269     -87.3105  \n",
       "LLR                            1.37795     0.192801  \n",
       "fraction_spatial_variance     0.991998     0.506044  \n",
       "pval                          0.240451     0.660595  \n",
       "qval                          0.977568     0.977568  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results.head().T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g</th>\n",
       "      <th>l</th>\n",
       "      <th>qval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>720</th>\n",
       "      <td>Cck</td>\n",
       "      <td>1.135190</td>\n",
       "      <td>8.802861e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865</th>\n",
       "      <td>Ptn</td>\n",
       "      <td>1.907609</td>\n",
       "      <td>8.162537e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>530</th>\n",
       "      <td>Prokr2</td>\n",
       "      <td>1.135190</td>\n",
       "      <td>1.916110e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>505</th>\n",
       "      <td>Nr2f2</td>\n",
       "      <td>1.135190</td>\n",
       "      <td>4.790035e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495</th>\n",
       "      <td>Frzb</td>\n",
       "      <td>1.135190</td>\n",
       "      <td>1.317798e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>Olfr635</td>\n",
       "      <td>0.675535</td>\n",
       "      <td>1.963151e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>437</th>\n",
       "      <td>Map1b</td>\n",
       "      <td>1.135190</td>\n",
       "      <td>4.955250e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>615</th>\n",
       "      <td>Agt</td>\n",
       "      <td>1.135190</td>\n",
       "      <td>6.470005e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>Cpne4</td>\n",
       "      <td>1.135190</td>\n",
       "      <td>7.150909e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>Sncb</td>\n",
       "      <td>1.135190</td>\n",
       "      <td>8.444712e-02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           g         l          qval\n",
       "720      Cck  1.135190  8.802861e-07\n",
       "865      Ptn  1.907609  8.162537e-05\n",
       "530   Prokr2  1.135190  1.916110e-03\n",
       "505    Nr2f2  1.135190  4.790035e-03\n",
       "495     Frzb  1.135190  1.317798e-02\n",
       "180  Olfr635  0.675535  1.963151e-02\n",
       "437    Map1b  1.135190  4.955250e-02\n",
       "615      Agt  1.135190  6.470005e-02\n",
       "351    Cpne4  1.135190  7.150909e-02\n",
       "397     Sncb  1.135190  8.444712e-02"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results.sort_values('qval').head(10)[['g', 'l', 'qval']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We detected a few spatially differentially expressed genes, _Cck_ and _Ptn_ for example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
